libgit/autogit_manager.go

```diff
@@ -0,0 +1,404 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libgit
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/eapache/channels"
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/kbfs/libfs"
+	"github.com/keybase/kbfs/libkbfs"
+	billy "gopkg.in/src-d/go-billy.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+type resetReq struct {
+	srcTLF     *libkbfs.TlfHandle
+	srcRepo    string
+	branchName string
+	dstTLF     *libkbfs.TlfHandle
+	dstDir     string
+	doneCh     chan struct{}
+}
+
+func (r resetReq) dstPath() string {
+	return path.Clean(path.Join(
+		r.dstTLF.GetCanonicalPath(), r.dstDir, r.srcRepo, r.branchName))
+}
+
+const (
+	cloneFileName = "CLONING"
+
+	// Debug tag ID for an individual autogit operation
+	ctxOpID = "AGM"
+)
+
+type ctxTagKey int
+
+const (
+	ctxIDKey ctxTagKey = iota
+)
+
+func autogitLockName(srcRepo string) string {
+	return fmt.Sprintf(".%s.lock", srcRepo)
+}
+
+type getNewConfigFn func(context.Context) (
+	context.Context, libkbfs.Config, string, error)
+
+// AutogitManager can clone and pull source git repos into a
+// destination folder, potentially across different TLFs.  New
+// requests for an operation in a destination folder are blocked by
```

Comment:
So "destination folder" should be "destination repo+branch" here per `resetReq.dstPath()` implementation.

Reply:
Also, this means reset requests for the same repo with different branch names won't block each other or get rolled into one. Is that intended? Does it affect user-facing "checked-out" branch, or is this just to get all necessary objects written?

Reply:
The intent is that the caller will use any non-master branch name in the `dstDir` passed to `Clone`/`Pull`.  I guess I should just take the branch name out of the id.  I'll write a comment to clarify.